Download API (#858)
MicrosoftEdge/WebView2Feedback: specs/CustomDownload.md

    For example, this section is a place to explain why you're adding this
    API rather than modifying an existing API.

    For example, this is a place to provide a brief explanation of some dependent
    area, just explanation enough to understand this new API, rather than telling
    the reader "go read 100 pages of background information posted at ...".
-->
There have been multiple requests by WebView Developers to be able to customize/control the download experience within WebView2.

Some of the requests we've heard, and want to solve with this API are:
- Developers want to be able to disable downloads for the entire environment.
- Developers want to be able to disable downloads on a per-case basis. For example, based on the content type, or user.
- Developers want to have access to downloaded file’s metadata.
- Developers want to modify the download path.
- Developers don’t want to see edge download UI if download is cancelled.

In this document we describe the updated API. We'd appreciate your feedback.

# Description

Our proposed solution is to give developers an API to control the download experience.
We will do this by exposing a DownloadStarting event.

> The DownloadStarting event will be raised when a download has begun.

Reviewer: I'm curious. There are hosts which only allows files to be downloaded ONCE, no resume. If a developer cancels the Edge download, takes the URL and downloads it himself, would that work? Or will cancelling the download makes Edge informing the website, that the download has been aborted?

Reviewer 2: Thanks for the question, do you mind expanding on your scenario? Is this a common scenario for you or an edge case? The way we intend for it to work is: a downloadStarting event triggers when the download begins, it would NOT cancel the download immediately. During this time, you can use the url and download metadata, to do any sort of confirmation, and block it from there. If you don't choose to block it, the download can resume as intended. Although, if you do infact block the download, expect it to work the same as cancelling a download in edge.

Reviewer: I think it's an edge case as >90% of all downloads on the web allows several attempts. If I understand you right, then file will be downloaded in background anyway and the developer only have the choice to abort this download. If thats true, what happens if the file is so small, lets say 1kb, and its finishing the download before the developer has taken any action on DownloadStarting event. Is the file then present on the user's hard disk?

> The host can then choose to cancel a download, change the save path, and hide the default download dialog.

Reviewer: If we're hiding the default UI download dialog of Edge. How we're informed about the download progress then?! Let's assume we have a 4GB file to download and we would like to hide the Edge download UI and display our own instead. How can we tell the user how much bytes have been downloaded already? From what I see we cannot, the download will be executed in background and nobody but Edge knows the progress and when its done.

Reviewer 3: I feel the same. For a complete experience we would also need a DownloadProgress event with the amount of bytes received up to that point.

Second, but I'm not that familiar with `Windows.Foundation.Deferral`, it reads that it only signals on download completed. How can be distinguished if the download completed succesfully, or failed somewhere?

If it failed we would also need a reason. Connection lost, out of diskspace, ...

Reviewer 2: Thanks @symbai and @frankdekker, both of you are correct, just blocking the download dock would leave a question mark about download progress. We intend to release a download progress API, although we are still early in the api drafting process, which is why it is not included in this PR. That said, we have heard customers request a way to simply disable the UI, which this API would solve.

Frank, to your second question: would a download completed event be useful? This is a valid request, and I can add it to our backlog. And to your 3rd question, the oppurtunity to cancel the download programmatically would happen when you intercept the downloadStarting event, do you expect to cancel the download otherwise?

Reviewer 3: @jasonstephen15 
I'm reasoning from the perspective that I can build a custom download GUI. To present the user a fully functional gui I need to know the following events:

- DownloadStartingEvent - where I can prompt the user if he wants to open the download it or save it at a specified location.
- DownloadProgressEvent - where I can show a download progress indicator somewhere in my application.
   -  this should contain info about the bytesReceived and pref about the transfer speed.
- DownloadCompletedEvent - where i can tell the user that the download finished
   - this should contain info about if it completed or failed.

Additionally a user could accept the download, but if it takes too long, still wants to choose to abort the download. The API should provide a mechanisme to abort the download. I assume cancelling the DownloadStartingEvent only allows the ability to cancel it at the start. For complete functionality we should also be able to cancel it _after_ the DownloadStartingEvent.

As an example the API that I'm used to in a Java environment:

DownloadStartingEvent contains a DownloadItem object and contains the following:
- event DownloadProgress
- event DownloadCompleted
- event DownloadFailed
- void Cancel()

As developer this provides me with the possibility to control every aspect of a single item download.

}

/// Event args for the `DownloadStarting` event.
[uuid(c7e95e2f-f789-44e6-b372-3141469240e4), object, pointer_default(unique)]
interface ICoreWebView2DownloadStartingEventArgs : IUnknown
{
  /// The URI of the download.
  [propget] HRESULT Uri([out, retval] LPWSTR* uri);

  /// The Content-Disposition header value from HTTP response.
  [propget] HRESULT ContentDisposition([out, retval] LPWSTR* contentDisposition);

  /// MIME type of the downloaded content.
  [propget] HRESULT MimeType([out, retval] LPWSTR* mimeType);

  /// The size of the download in total number of expected bytes.
  [propget] HRESULT DownloadSizeInBytes([out, retval] UINT64* downloadSizeInBytes);

  /// The host may set this flag to cancel the download. If canceled, the
  /// download save dialog will not be displayed regardless of the
  /// ShouldDisplayDefaultDownloadDialog property.
  [propget] HRESULT Cancel([out, retval] BOOL* cancel);

  /// Sets the `Cancel` property.
  [propput] HRESULT Cancel([in] BOOL cancel);

  /// The path to the file. If setting the path, the host should ensure that it
  /// is an absolute path, including the file name. If the directory does not
  /// exist, it will be created.

>   [propget] HRESULT ResultFilePath([out, retval] LPWSTR* resultFilePath);

Reviewer 3: What happens when this path points to an existing file? Will the file be overwritten or will it receive a suffix?

                m_completeDeferredDownloadEvent = [showDialog, deferral] {
                    showDialog();
                    CHECK_FAILURE(deferral->Complete());
                };

                return S_OK;
            })
            .Get(),
        &m_DownloadStartingToken));
```
## .NET/ WinRT
```c#
webView.CoreWebView2.DownloadStarting += delegate (object sender, CoreWebView2DownloadStartingEventArgs args)
{
    CoreWebView2Deferral deferral = args.GetDeferral();
    System.Threading.SynchronizationContext.Current.Post((_) =>
    {
        using (deferral)
        {
        
            if (_coreWebView2Settings == null)
            {
                _coreWebView2Settings = webView.CoreWebView2.Settings;
            }
            _coreWebView2Settings.ShouldDisplayDefaultDownloadDialog = false;

            var dialog = new TextInputDialog(
                title: "Download Starting",
                description: "Enter new result file path or select OK to keep default path. Select cancel to cancel the download.",
                defaultInput: args.ResultFilePath);
            if (dialog.ShowDialog() == true)
            {

>             args.ResultFilePath = dialog.Input.Text;

Reviewer 4: indenting

            }
            else
            {
            args.Cancel = true;
            }
        }
    }, null);
};

> ```

Reviewer 4: We're missing a sample for the new functionality / scenario covered by the doc - creating a new download UI. This could just be in C# / WinRT. No need to be fully working but if you have the IDL compiling it does make it easier to write the code in VS and ensure the code should be able to build.

    Include every new or modified type but use // ... to remove any methods,
    properties, or events that are unchanged.

    (GitHub's markdown syntax formatter does not (yet) know about MIDL3, so
    use ```c# instead even when writing MIDL3.)

    Example:

    ```
    /// Event args for the NewWindowRequested event. The event is fired when content
    /// inside webview requested to open a new window (through window.open() and so on.)
    [uuid(34acb11c-fc37-4418-9132-f9c21d1eafb9), object, pointer_default(unique)]
    interface ICoreWebView2NewWindowRequestedEventArgs : IUnknown
    {
        // ...

        /// Window features specified by the window.open call.
        /// These features can be considered for positioning and sizing of
        /// new webview windows.
        [propget] HRESULT WindowFeatures([out, retval] ICoreWebView2WindowFeatures** windowFeatures);
    }
    ```

    ```c# (but really MIDL3)
    public class CoreWebView2NewWindowRequestedEventArgs
    {
        // ...

	       public CoreWebView2WindowFeatures WindowFeatures { get; }
    }
    ```
-->
## Win32 C++
```c#
// Enums and structs

typedef enum COREWEBVIEW2_DOWNLOAD_DANGER_TYPE {

>   // The download is safe.

Author: Removing danger kind from API for now

  DANGER_TYPE_NONE,
  // The download file is dangerous.
  DANGER_TYPE_FILE,
  // The URI leads to a malicious file download.
  DANGER_TYPE_URI,
  // The content of the file is dangerous.
  DANGER_TYPE_CONTENT,
  // Not enough data to know whether download is dangerous.
  DANGER_TYPE_UNCOMMON,
  // The host is known to serve dangerous content.
  DANGER_TYPE_HOST,
  // The download is an application or extension that modifies browser or
  // computer settings.
  DANGER_TYPE_POTENTIALLY_UNWANTED,
} COREWEBVIEW2_DOWNLOAD_DANGER_TYPE;

typedef enum COREWEBVIEW2_DOWNLOAD_STATE {
  /// The download is in progress.
  DOWNLOAD_IN_PROGRESS,
  /// The download has been interrupted.
  DOWNLOAD_INTERRUPTED,
  /// The download is complete.

>   DOWNLOAD_COMPLETE,

Reviewer 4: Is there a distinction between paused and failed?

Author: Yes, an interrupted download is a failed download, although in some cases an interrupted download can be resumed. A paused download (paused by user) is in the `DOWNLOAD_IN_PROGRESS` state, not `DOWNLOAD_INTERRUPTED`, and it can always be resumed. Hopefully this is clear in the comments now.

Reviewer 4: If the download is paused by the user and the state doesn't change, how will the app know they should call Resume to resume the download?

Author: The app can handle the user pausing and resuming a download through the Pause and Resume methods on the download. The properties `CanResume` and `IsPaused` are also available. Is that good enough?

Reviewer 4: Ideally there's one set of code to deal with the download stopping/interruptions/pausing in the state change event handler rather than also having additional code to handle the user pausing the download and having to separately track that info and having separate code for that.

  /// download save dialog will not be displayed regardless of the
  /// ShouldDisplayDefaultDownloadDialog property.
  [propget] HRESULT Cancel([out, retval] BOOL* cancel);

  /// Sets the `Cancel` property.
  [propput] HRESULT Cancel([in] BOOL cancel);

  /// The path to the file. If setting the path, the host should ensure that it
  /// is an absolute path, including the file name. If the directory does not
  /// exist, it will be created.
  [propget] HRESULT ResultFilePath([out, retval] LPWSTR* resultFilePath);

  /// Sets the `ResultFilePath` property.
  [propput] HRESULT ResultFilePath([in] LPCWSTR resultFilePath);

  /// Returns an `ICoreWebView2Deferral` object.  Use this operation to
  /// complete the event at a later time.
  HRESULT GetDeferral([out, retval] ICoreWebView2Deferral** deferral);
}

/// Represents a download process. Gives access to the download's metadata
/// and supports a user canceling, pausing, or resuming the download.
[uuid(0C4A07B1-B610-459F-A574-EC253D5EC40D), object, pointer_default(unique)]
interface ICoreWebView2DownloadItem : IUnknown
{
  /// The unique ID of the download.

>   [propget] HRESULT Id([out, retval] UINT32* id);

Reviewer 4: What would we do with this Id property? Can we use it anywhere?

  /// The URI of the download.
  [propget] HRESULT Uri([out, retval] LPWSTR* uri);

  /// The Content-Disposition header value from HTTP response.
  [propget] HRESULT ContentDisposition([out, retval] LPWSTR* contentDisposition);

  /// MIME type of the downloaded content.
  [propget] HRESULT MimeType([out, retval] LPWSTR* mimeType);

  /// The size of the download in total number of expected bytes.

>   [propget] HRESULT DownloadSizeInBytes([out, retval] UINT64* downloadSizeInBytes);

Author: It is the download size before file compression. Should I add Expected?

>   [propget] HRESULT DownloadSizeInBytes([out, retval] UINT64* downloadSizeInBytes);

Reviewer 4: What do you mean by file compression? Do you mean the encodings applied to the HTTP body as its is transferred? I'm not worried about that. I think this has to be the expected size based on the initial HTTP response's headers and that it could be missing or incorrect compared to what we actually end up getting? If so then add Expected. Does this value change as the download goes on or completes?

>   [propget] HRESULT DownloadSizeInBytes([out, retval] UINT64* downloadSizeInBytes);

Author: Yes you're right, it comes from content-length header. Added Expected

  /// The danger type of the download. The danger type can be none, file, content,
  /// uncommon, host, or potentially unwanted. See `COREWEBVIEW2_DOWNLOAD_DANGER_TYPE`
  /// for more details.
  [propget] HRESULT DangerType([out, retval] COREWEBVIEW2_DOWNLOAD_DANGER_TYPE* dangerType);

  /// The state of the download. A download can be in progress, interrupted, or
  /// complete.
  [propget] HRESULT State([out, retval] COREWEBVIEW2_DOWNLOAD_STATE* downloadState);

  /// The reason why a download was interrupted.
  [propget] HRESULT InterruptReason([out, retval] LPWSTR* interruptReason);

  /// Cancels the download. If canceled, the default download dialog will show that the
  /// download was canceled. Host should set `Cancel` from `ICoreWebView2SDownloadStartingEventArgs`
  /// if cancellation should be hidden from default download dialog.
  HRESULT Cancel();

  /// Pauses the download. If paused, the default download dialog will
  /// show that the download is paused. No effect if download is already paused.
  HRESULT Pause();

  /// Resumes a paused download. No effect if download is not paused.
  HRESULT Resume();

> }

Reviewer 4: Don't we need an event on the Download class to say when the state changes or the receivedbytes and so on properties change?

    For example, this is a place to provide a brief explanation of some dependent
    area, just explanation enough to understand this new API, rather than telling
    the reader "go read 100 pages of background information posted at ...".
-->
We are redesigning the download experience by providing a DownloadStarting event, and exposing a Download object with all associated download metadata. With this you will be able to block downloads, save to a different path, and have access to the required metadata to build your own download UI.

In this document we describe the updated API. We'd appreciate your feedback.

# Description

There are 3 parts to this API.

1. DownloadStarting [Event] - used to intercept a download.
    - Download (link to the download object below)
    - Cancel
    - Result file path (set)
2. Download [Object]: This will give you all the metadata that you might need to block a download, and build UI.
    - Uri
    - Mime type
    - Content disposition
    - Download size in bytes
    - Result file path (get)
    - Progress size in bytes
    - Estimated end time
    - State (in progress, completed, interrupted)
    - Pause/resume/cancel

>     - Deferral

Reviewer 4: Deferral is on the event args not the download class, correct?

Author: Correct

Reviewer 4: Great! Please move to event

    - Events:
        - DownloadStateChanged
            - InterruptReason
        - DownloadProgressSizeInBytesChanged
        - DownloadEstimatedEndTimeChanged
3. WebView Setting: Default Download Dialog - used to disable the default download UI (bottom dock)

> We believe these 3 parts of a redesigned Download API, should enable your app have a customizable & complete download experience.

Reviewer 4: ```suggestion
We believe these 3 parts of a redesigned Download API, should enable your app to have a customizable & complete download experience.
```

    }
    ```

    ```cpp
    void SampleClass::SampleMethod()
    {
        winrt::com_ptr<ICoreWebView2> webview2 = ...
    }
    ```

    As an example of this section, see the Examples section for the PasswordBox
    control (https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples).
-->
## Win32 C++
```cpp
// Register a handler for the `DownloadStarting` event.
// This example hides the default download dialog and shows a dialog box instead.
// The dialog box displays the default save path and allows the user to specify a different path.
// Selecting `OK` will save the download to the chosen path.
// Selecting `CANCEL` will cancel the download.
CHECK_FAILURE(m_webView->add_DownloadStarting(
    Callback<ICoreWebView2DownloadStartingEventHandler>(
        [this](
            ICoreWebView2_3* sender,
            ICoreWebView2DownloadStartingEventArgs* args) -> HRESULT {
            auto showDialog = [this, args] {

                // Hide the default download dialog.
                wil::com_ptr<ICoreWebView2Settings> m_settings;
                CHECK_FAILURE(m_webView->get_Settings(&m_settings));

>                 CHECK_FAILURE(

Reviewer 4: The ShouldDisplayDefaultDownloadDialog setting applies to the entire webview2. You probably want to set this once outside of your event handler correct not again and again inside your event handler? Also not sure if its too late to set this inside the event handler?

                    m_settings->put_ShouldDisplayDefaultDownloadDialog(FALSE));

                wil::com_ptr<ICoreWebView2Download> download;

>                 CHECK_FAILURE(args->Download(&download));

Reviewer 4: Should be get_Download right?

                if (dialog.confirmed)
                {
                    // If user selects `OK`, the download will complete normally.
                    // Result file path will be updated if a new one was provided.
                    CHECK_FAILURE(args->put_ResultFilePath(dialog.input.c_str()));
                    UpdateProgress(download.get());
                }
                else
                {
                    // If user selects `Cancel`, the download will be canceled.
                    CHECK_FAILURE(args->put_Cancel(TRUE));
                }
            };

            wil::com_ptr<ICoreWebView2Deferral> deferral;
            CHECK_FAILURE(args->GetDeferral(&deferral));

            // This function can be called to show the download dialog and
            // complete the event at a later time, allowing the developer to
            // perform async work before the event completes.
            m_completeDeferredDownloadEvent = [showDialog, deferral] {
                showDialog();
                CHECK_FAILURE(deferral->Complete());
            };

            return S_OK;
        })
    .Get(),
&m_downloadStartingToken));

// Update download progress

> // Register a handler for the `DownloadProgressSizeInBytesChanged` event.

Reviewer 4: Is this supposed to be the UpdateProgress method used above? Or where does this download variable come from? Is this part of the above sample snippet, or is this its own sample snippet? If so, you'd want to have download starting event handler that does this work below. If not, then good to hook it up to the above with appropriate method call or whatever

Author: Yes it's meant to be the UpdateProgress method

Reviewer 4: Yes

                break;
            case DOWNLOAD_INTERRUPTED:
                // Here developer can take different actions based on `args->InterruptReason`.
                MessageBox(nullptr, L"Connection with file host was broken.", nullptr, MB_OK);
                CHECK_FAILURE(download->get_CanResume(&canResume));
                if (canResume)
                {
                    download->Resume();
                }
                else
                {
                    CompleteDownload(download);
                }
                break;
            case DOWNLOAD_COMPLETE:
                CompleteDownload(download);
                break;
            }
            return S_OK;
        })
    .Get(),
&m_downloadStateChangedToken));
```
## .NET/ WinRT
```c#
webView.CoreWebView2.DownloadStarting += delegate (object sender, CoreWebView2DownloadStartingEventArgs args)
{
    CoreWebView2Deferral deferral = args.GetDeferral();
    System.Threading.SynchronizationContext.Current.Post((_) =>
    {
        using (deferral)
        {
            if (_coreWebView2Settings == null)
            {
                _coreWebView2Settings = webView.CoreWebView2.Settings;
            }

>             _coreWebView2Settings.ShouldDisplayDefaultDownloadDialog = false;

Reviewer 4: Same note about doing this in the event handler vs once outside the event handler

                defaultInput: args.ResultFilePath);
            if (dialog.ShowDialog() == true)
            {
                args.ResultFilePath = dialog.Input.Text;
                UpdateProgress(args.Download());
            }
            else
            {
                args.Cancel = true;
            }
        }
    }, null);
};

// Update download progress
void UpdateProgress(CoreWebView2Download download)
{
    download.DownloadProgressSizeInBytesChanged += delegate (object sender, Object e)
    {
        // Here developer can update download dialog to show progress of a download using
        // `download.DownloadProgressSizeInBytes` and `download.DownloadSizeInBytes`
    };

    download.DownloadStateChanged += delegate (object sender, CoreWebView2DownloadStateChangedEventArgs args)
    {
        switch (download.State)
        {
        case CoreWebView2DownloadState.DownloadInProgress:
            break;
        case CoreWebView2DownloadState.DownloadInterrupted:
            // Here developer can take different actions based on `args.InterruptReason`.

>             MessageBox.Show(this, "Connecton with file host broken.", "Download Interrupted");

Reviewer 4: Issue of running messagebox during event handler

Reviewer 4: and for the C++ sample

Author: Right. I'll just remove it?

Reviewer 4: Remove or replace with method for which you don't show the definition.

>             if (download.CanResume == true)

Reviewer 4: ```suggestion
            if (download.CanResume)
```

            {

>                 download.Resume();

Reviewer 4: We probably don't want to auto resume. If we're interrupted for a reason that must be resolved and we never try to resolve it or tell the user who can try to resolve, then we're just endlessly getting interrupted & resuming. We probably don't want to tell end devs to do it this way.

    Include every new or modified type but use // ... to remove any methods,
    properties, or events that are unchanged.

    (GitHub's markdown syntax formatter does not (yet) know about MIDL3, so
    use ```c# instead even when writing MIDL3.)

    Example:

    ```
    /// Event args for the NewWindowRequested event. The event is fired when content
    /// inside webview requested to open a new window (through window.open() and so on.)
    [uuid(34acb11c-fc37-4418-9132-f9c21d1eafb9), object, pointer_default(unique)]
    interface ICoreWebView2NewWindowRequestedEventArgs : IUnknown
    {
        // ...

        /// Window features specified by the window.open call.
        /// These features can be considered for positioning and sizing of
        /// new webview windows.
        [propget] HRESULT WindowFeatures([out, retval] ICoreWebView2WindowFeatures** windowFeatures);
    }
    ```

    ```c# (but really MIDL3)
    public class CoreWebView2NewWindowRequestedEventArgs
    {
        // ...

	       public CoreWebView2WindowFeatures WindowFeatures { get; }
    }
    ```
-->
## Win32 C++
```c#
// Enums and structs

typedef enum COREWEBVIEW2_DOWNLOAD_STATE {

>   /// The download is in progress. Includes downloads paused by user.

Reviewer 4: Paused downloads are in progress? They don't become interrupted?

Author: No they are not interrupted, they remain in progress :/ Maybe we want to use a different word? Failed?

Reviewer 4: I'm confused. In what way are they in progress? Does the download continue to occur in the background? Or do you mean the internal chromium code says the state is in progress? Can our code not distinguish between the download is progressing as normal and the user has paused the download?

Author: The internal chromium code says the state is in progress when the download is paused, but we can check using IsPaused on the chromium download item

Reviewer 4: But does the chromium code call one of our callbacks to let us know that a download is paused so we have an opportunity to call IsPaused? Is there something significantly different between the download being paused this way and the download being interrupted in a different way that we wouldn't want to represent them using the same state/enum and so on?

>   DOWNLOAD_IN_PROGRESS,

Reviewer 4: In COM our enums all start with the name of the enum like COREWEBVIEW2_DOWNLOAD_STATE_*. In .NET and WinRT this isn't the case and our generated code will remove that prefix along with changing the casing and removing _.

So here it could be COREWEBVIEW2_DOWNLOAD_STATE_IN_PROGRESS which would become InProgress in .NET & WinRT.

  /// The connection with the file host was broken. The `InterruptReason` property
  /// can be accessed from `ICoreWebView2DownloadStateChangedEventArgs`. See
  /// `COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON` for descriptions of kinds of
  /// interrupt reasons. Host can check whether an interrupted download can be

>   /// resumed with the `CanResume` method on the `ICoreWebView2Download`. Once

Reviewer 4: ```suggestion
  /// resumed with the `CanResume` property on the `ICoreWebView2Download`. Once
```

  /// resumed, a download is in the `DOWNLOAD_IN_PROGRESS` state.
  DOWNLOAD_INTERRUPTED,
  /// The download completed successfully.

>   DOWNLOAD_COMPLETE,

Reviewer 4: Can this be COMPLETE**D**?

} COREWEBVIEW2_DOWNLOAD_STATE;

> typedef enum COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON {

Reviewer 4: Same note about prefix on all the enum names: COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_*.

  NETWORK_SERVER_DOWN,
  /// Network request invalid because original or redirected URI is invalid, has
  /// an unsupported scheme, or is disallowed by policy.
  NETWORK_INVALID_REQUEST,

  /// Server responses
  /// Generic server error.
  SERVER_FAILED,
  /// Server does not support range requests.
  SERVER_NO_RANGE,
  /// Server does not have the requested data.
  SERVER_BAD_CONTENT,
  /// Server did not authorize access to resource.
  SERVER_UNAUTHORIZED,
  /// Server certificate problem.
  SERVER_CERT_PROBLEM,
  /// Server access forbidden.
  SERVER_FORBIDDEN,
  /// Unexpected server response. Responding server may not be intended server.
  SERVER_UNREACHABLE,
  /// Server sent fewer bytes than the content-length header. Content-length
  /// header may be invalid or connection may have closed. Download is treated
  /// as complete unless there are strong validators present to interrupt the
  /// download.
  SERVER_CONTENT_LENGTH_MISMATCH,
  /// Unexpected cross-origin redirect.
  SERVER_CROSS_ORIGIN_REDIRECT,

  /// User input
  /// User canceled the download.
  USER_CANCELED,
  /// User shut down the WebView.
  USER_SHUTDOWN,

  /// WebView crashed.

>   CRASH,

Reviewer 4: ```suggestion
  DOWNLOAD_PROCESS_CRASHED,
```

} COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON;

[uuid(9aab8652-d89f-408d-8b2c-1ade3ab51d6d), object, pointer_default(unique)]

> interface ICoreWebView2Settings : IUnknown {

Reviewer 4: ```suggestion
interface ICoreWebView2Settings2 : ICoreWebView2Settings {
```

interface ICoreWebView2DownloadStartingEventHandler : IUnknown
{
  /// Provides the event args for the corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2_3* sender,
      [in] ICoreWebView2DownloadStartingEventArgs* args);
}

/// Event args for the `DownloadStarting` event.
[uuid(c7e95e2f-f789-44e6-b372-3141469240e4), object, pointer_default(unique)]
interface ICoreWebView2DownloadStartingEventArgs : IUnknown
{
  /// Returns the `ICoreWebView2Download` for the download that has started.
  HRESULT Download([out, retval] ICoreWebView2Download** download);

  /// The host may set this flag to cancel the download. If canceled, the
  /// download save dialog will not be displayed regardless of the
  /// ShouldDisplayDefaultDownloadDialog property.
  [propget] HRESULT Cancel([out, retval] BOOL* cancel);

  /// Sets the `Cancel` property.
  [propput] HRESULT Cancel([in] BOOL cancel);

  /// The path to the file. If setting the path, the host should ensure that it
  /// is an absolute path, including the file name. If the directory does not
  /// exist, it will be created. If the path points to an existing file, the

>   /// file name will receive a suffix.

Reviewer 4: ```suggestion
  /// actual file name used will have an ` (N)` suffix appended.
```

[uuid(D98F73CE-D94D-4576-A1E0-EB8989A9DBEF), object, pointer_default(unique)]
interface ICoreWebView2DownloadProgressSizeInBytesChangedEventHandler : IUnknown
{
  /// Provides the event args for the corresponding event. No event args exist
  /// and the `args` parameter is set to `null`.
  HRESULT Invoke(
      [in] ICoreWebView2Download* sender, [in] IUnknown* args);
}

/// Implements the interface to receive `DownloadEstimatedEndTimeChanged` event. Use the
/// `ICoreWebView2Download.EstimatedEndTime` property to get the new estimated
/// end time.

[uuid(D46CD703-958B-4100-9F84-4E6C2FC32D57), object, pointer_default(unique)]
interface ICoreWebView2DownloadEstimatedEndTimeChangedEventHandler : IUnknown
{
  /// Provides the event args for the corresponding event. No event args exist
  /// and the `args` parameter is set to `null`.
  HRESULT Invoke(
      [in] ICoreWebView2Download* sender, [in] IUnknown* args);
}

/// Implements the interface to receive `DownloadStateChanged` event. Use the
/// `ICoreWebView2Download.State` property to get the current state, which can
/// be in progress, interrupted, or complete. Find the `InterruptReason` property

> /// on 'ICoreWebView2DownloadStateChangedEventArgs`.

Reviewer 4: ```suggestion
/// on `ICoreWebView2DownloadStateChangedEventArgs`.
```

        FileHashMismatch = 11,
        FileSameAsSource = 12,
        NetworkFailed = 13,
        NetworkTimeout = 14,
        NetworkDisconnected = 15,
        NetworkServerDown = 16,
        NetworkInvalidRequest = 17,
        ServerFailed = 18,
        ServerNoRange = 19,
        ServerBadContent = 20,
        ServerUnauthorized = 21,
        ServerCertProblem = 22,
        ServerForbidden = 23,
        ServerUnreachable = 24,
        ServerContentLengthMismatch = 25,
        ServerCrossOriginRedirect = 26,
        UserCanceled = 27,
        UserShutdown = 28,
        Crash = 29,
    };

    runtimeclass CoreWebView2Settings
    {
        // CoreWebView2Settings
        // There are other settings that we are not showing
        Boolean ShouldDisplayDefaultDownloadDialog { get; set; };
    }

    runtimeclass CoreWebView2DownloadStateChangedEventArgs
    {
        // CoreWebView2DownloadStateChangedEventArgs
        CoreWebView2DownloadInterruptReason InterruptReason { get; };
    }

    runtimeclass CoreWebView2DownloadStartingEventArgs
    {
        // CoreWebView2DownloadStartingEventArgs
        Boolean Cancel { get; set; };

        String ResultFilePath { get; set; };

>         CoreWebView2Download Download();

Reviewer 4: this should be a download property not method

        Windows.Foundation.Deferral GetDeferral();
    }

    runtimeclass CoreWebView2
    {
        // There are other API in this interface that we are not showing
        event Windows.Foundation.TypedEventHandler<CoreWebView2, CoreWebView2DownloadStartingEventArgs> DownloadStarting;
    }

    runtimeclass CoreWebView2Download
    {
        // CoreWebView2Download
        String Uri { get; };

        String ContentDisposition { get; };

        String MimeType { get; };

        UInt64 DownloadSizeInBytes { get; };

        UInt64 DownloadProgressSizeInBytes { get; };

>         String EstimatedEndTime { get; };

Reviewer 4: This should probably be a winrt DateTime object instead of a string?

    replaces or supplements existing functionality.

    ```c#
    void SampleMethod()
    {
        var show = new AnExampleOf();
        show.SomeMembers = AndWhyItMight(be, interesting)
    }
    ```

    ```cpp
    void SampleClass::SampleMethod()
    {
        winrt::com_ptr<ICoreWebView2> webview2 = ...
    }
    ```

    ## SecondFeatureName

    Feature explanation text goes here, including why an app would use it, how it
    replaces or supplements existing functionality.

    ```c#
    void SampleMethod()
    {
        var show = new AnExampleOf();
        show.SomeMembers = AndWhyItMight(be, interesting)
    }
    ```

    ```cpp
    void SampleClass::SampleMethod()
    {
        winrt::com_ptr<ICoreWebView2> webview2 = ...
    }
    ```

    As an example of this section, see the Examples section for the PasswordBox
    control (https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples).
-->
## Win32 C++
```cpp
ScenarioCustomDownloadExperience::ScenarioCustomDownloadExperience(AppWindow* appWindow)
    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
{
    // Hide the default download dialog.

>     wil::com_ptr<ICoreWebView2Settings> m_settings;

Reviewer 4: m_ prefix is for members not local vars
```suggestion
    wil::com_ptr<ICoreWebView2Settings> settings;
```

    CHECK_FAILURE(m_webView->get_Settings(&m_settings));
    CHECK_FAILURE(m_settings->put_ShouldDisplayDefaultDownloadDialog(FALSE));

    // Register a handler for the `DownloadStarting` event.
    // This example hides the default download dialog and shows a dialog box instead.
    // The dialog box displays the default save path and allows the user to specify a different path.
    // Selecting `OK` will save the download to the chosen path.
    // Selecting `CANCEL` will cancel the download.
    CHECK_FAILURE(m_webView->add_DownloadStarting(
        Callback<ICoreWebView2DownloadStartingEventHandler>(
            [this](
                ICoreWebView2_3* sender,
                ICoreWebView2DownloadStartingEventArgs* args) -> HRESULT {

>                 auto showDialog = [this, args] {

Reviewer 4: { goes on its own line in our Microsoft coding style sample code (also prev line)

  HRESULT add_DownloadEstimatedEndTimeChanged(
    [in] ICoreWebView2DownloadEstimatedEndTimeChangedEventHandler* eventHandler,
    [out] EventRegistrationToken* token);

  /// Remove an event handler previously added with `add_DownloadEstimatedEndTimeChanged`.
  HRESULT remove_DownloadEstimatedEndTimeChanged(
      [in] EventRegistrationToken token);

  /// Add an event handler for the `DownloadStateChanged` event.
  HRESULT add_DownloadStateChanged(
    [in] ICoreWebView2DownloadStateChangedEventHandler* eventHandler,
    [out] EventRegistrationToken* token);

  /// Remove an event handler previously added with `add_DownloadStateChanged`.
  HRESULT remove_DownloadStateChanged(
      [in] EventRegistrationToken token);

  /// The URI of the download.
  [propget] HRESULT Uri([out, retval] LPWSTR* uri);

  /// The Content-Disposition header value from HTTP response.
  [propget] HRESULT ContentDisposition([out, retval] LPWSTR* contentDisposition);

  /// MIME type of the downloaded content.
  [propget] HRESULT MimeType([out, retval] LPWSTR* mimeType);

  /// The expected size of the download in total number of bytes based on the

>   /// HTTP content-length header. Returns -1 if the size is unknown.

Reviewer 4: -1 but the type is unsigned.

Reviewer 4: I think the type should be unsigned though right? What is the maximum size of download supported? Or what is the type that chromium is using to tell us the size? Is -1 what chromium is giving us in this case?

Reviewer 4: Unresolving conversation since its still signed int.

Author: I made a mistake with unsigned int, the type in chromium is signed int for both received bytes and total bytes, and chromium gives us the -1 for total bytes if size is unknown.

# Background
<!-- TEMPLATE
    Use this section to provide background context for the new API(s)
    in this spec.

    This section and the appendix are the only sections that likely
    do not get copied into any official documentation, they're just an aid
    to reading this spec.

    If you're modifying an existing API, included a link here to the
    existing page(s) or spec documentation.

    For example, this section is a place to explain why you're adding this
    API rather than modifying an existing API.

    For example, this is a place to provide a brief explanation of some dependent
    area, just explanation enough to understand this new API, rather than telling
    the reader "go read 100 pages of background information posted at ...".
-->
We are redesigning the download experience by providing a DownloadStarting event, and exposing a Download object with all associated download metadata. With this you will be able to block downloads, save to a different path, and have access to the required metadata to build your own download UI.

In this document we describe the updated API. We'd appreciate your feedback.

# Description

> There are 3 parts to this API.

Reviewer 4: Rather than duplicate the API listing up here, it might be better to summarize the 3 parts like the following and remove the 1,2,3 listing below that sort of duplicates info from the API Details section at the bottom.
```suggestion
There are 3 parts to this API: the CoreWebView2.DownloadStarting event that can be used to find out about or cancel new downloads, the Download class available from the DownloadStarting event that has up to date information on a download as it progresses, and the CoreWebView2Settings.DefaultDownloadDialog property that can be used to disable the default download UI.
```

Reviewer 4: And also remove the 1,2,3 listing that is duplicating info from the API Details section at the bottom

    control (https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples).
-->
## Win32 C++
```cpp
ScenarioCustomDownloadExperience::ScenarioCustomDownloadExperience(AppWindow* appWindow)
    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
{
    // Hide the default download dialog.
    wil::com_ptr<ICoreWebView2Settings> settings;
    CHECK_FAILURE(m_webView->get_Settings(&settings));
    CHECK_FAILURE(settings->put_ShouldDisplayDefaultDownloadDialog(FALSE));

    // Register a handler for the `DownloadStarting` event.
    // This example hides the default download dialog and shows a dialog box instead.
    // The dialog box displays the default save path and allows the user to specify a different path.
    // Selecting `OK` will save the download to the chosen path.
    // Selecting `CANCEL` will cancel the download.
    CHECK_FAILURE(m_webView->add_DownloadStarting(
        Callback<ICoreWebView2DownloadStartingEventHandler>(
            [this](
                ICoreWebView2_3* sender,
                ICoreWebView2DownloadStartingEventArgs* args) -> HRESULT
            {

>                 auto showDialog = [this, args]

Reviewer 4: Please add a comment here explaining why we're doing this in a lambda
```suggestion
                // We avoid potential reentrancy from running a message loop in the download
                // starting event handler by showing our download dialog via this lambda run
                // asynchronously later outside of this event handler.
                auto showDialog = [this, args]
```

                                                L"Mime type: " + mimeType.get() + L"\r\n" +
                                                L"Expected Download size in bytes: " +
                                                std::to_wstring(expectedDownloadSizeInBytes) +
                                                L"\r\n";

                    TextInputDialog dialog(
                        m_appWindow->GetMainWindow(), L"Download Starting", prompt.c_str(),
                        description.c_str(), resultFilePath.get());
                    if (dialog.confirmed)
                    {
                        // If user selects `OK`, the download will complete normally.
                        // Result file path will be updated if a new one was provided.
                        CHECK_FAILURE(args->put_ResultFilePath(dialog.input.c_str()));
                        UpdateProgress(download.get());
                    }
                    else
                    {
                        // If user selects `Cancel`, the download will be canceled.
                        CHECK_FAILURE(args->put_Cancel(TRUE));
                    }
                };

>                 wil::com_ptr<ICoreWebView2Deferral> deferral;

Reviewer 4: ```suggestion
                // Obtain a deferral for the event so that the CoreWebView2
                // doesn't examine the properties we set on the event args until
                // after we call the Complete method asynchronously later.
                wil::com_ptr<ICoreWebView2Deferral> deferral;
```

    CHECK_FAILURE(download->add_DownloadProgressSizeInBytesChanged(
        Callback<ICoreWebView2DownloadProgressSizeInBytesChangedEventHandler>(
            [this](ICoreWebView2Download* download, IUnknown* args) -> HRESULT {
                // Here developer can update UI to show progress of a download using
                // `download->get_DownloadProgressSizeInBytes` and
                // `download->get_ExpectedDownloadSizeInBytes`.
                return S_OK;
            })
        .Get(),
    &m_downloadProgressSizeInBytesChangedToken));

    // Register a handler for the `DownloadStateChanged` event.
    CHECK_FAILURE(download->add_DownloadStateChanged(
        Callback<ICoreWebView2DownloadStateChangedEventHandler>(
            [this](ICoreWebView2Download* download,
            ICoreWebView2DownloadStateChangedEventArgs* args) -> HRESULT {
                COREWEBVIEW2_DOWNLOAD_STATE downloadState;
                CHECK_FAILURE(download->get_State(&downloadState));
                switch (downloadState)
                {
                case COREWEBVIEW2_DOWNLOAD_STATE_IN_PROGRESS:
                    break;
                case COREWEBVIEW2_DOWNLOAD_STATE_INTERRUPTED:

>                     // Here developer can take different actions based on `args->InterruptReason`.

Reviewer 4: ```suggestion
                    // Here developer can take different actions based on `args->InterruptReason`.
                    // For example, show an error message to the end user.
```

                    CompleteDownload(download);
                    break;
                case COREWEBVIEW2_DOWNLOAD_STATE_COMPLETED:
                    CompleteDownload(download);
                    break;
                }
                return S_OK;
            })
        .Get(),
    &m_downloadStateChangedToken));
}
```
## .NET/ WinRT
```c#
void DownloadStartingCmdExecuted(object target, ExecutedRoutedEventArgs e)
{
    if (_coreWebView2Settings == null)
    {
        _coreWebView2Settings = webView.CoreWebView2.Settings;
    }
    _coreWebView2Settings.ShouldDisplayDefaultDownloadDialog = false;

    webView.CoreWebView2.DownloadStarting += delegate (object sender, CoreWebView2DownloadStartingEventArgs args)
    {

>         CoreWebView2Deferral deferral = args.GetDeferral();

Reviewer 4: Please add similar comments from my suggestions in the COM sample to here as well

    }
    ```

    ```c# (but really MIDL3)
    public class CoreWebView2NewWindowRequestedEventArgs
    {
        // ...

	       public CoreWebView2WindowFeatures WindowFeatures { get; }
    }
    ```
-->
## Win32 C++
```c#
// Enums and structs

typedef enum COREWEBVIEW2_DOWNLOAD_STATE {
  /// The download is in progress.
  COREWEBVIEW2_DOWNLOAD_STATE_IN_PROGRESS,
  /// The connection with the file host was broken. The `InterruptReason` property
  /// can be accessed from `ICoreWebView2DownloadStateChangedEventArgs`. See
  /// `COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON` for descriptions of kinds of
  /// interrupt reasons. Host can check whether an interrupted download can be
  /// resumed with the `CanResume` property on the `ICoreWebView2Download`. Once
  /// resumed, a download is in the `COREWEBVIEW2_DOWNLOAD_STATE_IN_PROGRESS` state.
  COREWEBVIEW2_DOWNLOAD_STATE_INTERRUPTED,
  /// The download completed successfully.
  COREWEBVIEW2_DOWNLOAD_STATE_COMPLETED,
} COREWEBVIEW2_DOWNLOAD_STATE;

typedef enum COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON {
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NONE,

>   /// File errors

Reviewer 4: We don't have a way to do categories like this. It will just turn into documentation for the FILE_FAILED enum entry not a heading for all of the file enum entries.
```suggestion
```

  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_ACCESS_DENIED,
  /// Disk full.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_NO_SPACE,
  /// Result file path with file name is too long.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_NAME_TOO_LONG,
  /// File is too large for file system.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_TOO_LARGE,
  /// File contains a virus.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_VIRUS_INFECTED,
  /// File was in use, too many files opened, or out of memory.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_TRANSIENT_ERROR,
  /// File blocked by local policy.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_BLOCKED,
  /// Security check failed unexpectedly.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_SECURITY_CHECK_FAILED,
  /// Seeking past the end of a file in opening a file, as part of resuming an
  /// interrupted download.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_TOO_SHORT,
  /// Partial file did not match the expected hash.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_HASH_MISMATCH,
  /// Source and target of download were the same.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_SAME_AS_SOURCE,

>   /// Network errors

Reviewer 4: ```suggestion
```

  /// Generic network error.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_FAILED,
  /// Network operation timed out.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_TIMEOUT,
  /// Network connection lost.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_DISCONNECTED,
  /// Server has gone down.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_SERVER_DOWN,
  /// Network request invalid because original or redirected URI is invalid, has
  /// an unsupported scheme, or is disallowed by policy.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_INVALID_REQUEST,

>   /// Server responses

Reviewer 4: ```suggestion
```

  /// Generic server error.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_FAILED,
  /// Server does not support range requests.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_NO_RANGE,
  /// Server does not have the requested data.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_BAD_CONTENT,
  /// Server did not authorize access to resource.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_UNAUTHORIZED,
  /// Server certificate problem.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CERT_PROBLEM,
  /// Server access forbidden.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_FORBIDDEN,
  /// Unexpected server response. Responding server may not be intended server.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_UNREACHABLE,
  /// Server sent fewer bytes than the content-length header. Content-length
  /// header may be invalid or connection may have closed. Download is treated
  /// as complete unless there are strong validators present to interrupt the
  /// download.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CONTENT_LENGTH_MISMATCH,
  /// Unexpected cross-origin redirect.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CROSS_ORIGIN_REDIRECT,

>   /// User input

Reviewer 4: ```suggestion
```

  /// User canceled the download.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_CANCELED,
  /// User shut down the WebView.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_SHUTDOWN,
  /// User paused the download.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_PAUSED,

  /// WebView crashed.
  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_DOWNLOAD_PROCESS_CRASHED,
} COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON;

[uuid(9aab8652-d89f-408d-8b2c-1ade3ab51d6d), object, pointer_default(unique)]
interface ICoreWebView2Settings2 : ICoreWebView2Settings {

>   /// The host may set this flag to hide the default download dialog.

Reviewer 4: ```suggestion
  /// The host may set this flag to hide the default download dialog. The CoreWebView2.DownloadStarting event will continue to fire, and the download will progress as normal, there will just be no default download UI.
```

    [out] EventRegistrationToken* token);

  /// Remove an event handler previously added with `add_DownloadProgressSizeInBytesChanged`.
  HRESULT remove_DownloadProgressSizeInBytesChanged(
      [in] EventRegistrationToken token);

  /// Add an event handler for the `DownloadEstimatedEndTimeChanged` event.
  HRESULT add_DownloadEstimatedEndTimeChanged(
    [in] ICoreWebView2DownloadEstimatedEndTimeChangedEventHandler* eventHandler,
    [out] EventRegistrationToken* token);

  /// Remove an event handler previously added with `add_DownloadEstimatedEndTimeChanged`.
  HRESULT remove_DownloadEstimatedEndTimeChanged(
      [in] EventRegistrationToken token);

  /// Add an event handler for the `DownloadStateChanged` event.
  HRESULT add_DownloadStateChanged(
    [in] ICoreWebView2DownloadStateChangedEventHandler* eventHandler,
    [out] EventRegistrationToken* token);

  /// Remove an event handler previously added with `add_DownloadStateChanged`.
  HRESULT remove_DownloadStateChanged(
      [in] EventRegistrationToken token);

  /// The URI of the download.
  [propget] HRESULT Uri([out, retval] LPWSTR* uri);

>   /// The Content-Disposition header value from HTTP response.

Reviewer 4: ```suggestion
  /// The Content-Disposition header value from the download's HTTP response.
```

  [propget] HRESULT ContentDisposition([out, retval] LPWSTR* contentDisposition);

  /// MIME type of the downloaded content.
  [propget] HRESULT MimeType([out, retval] LPWSTR* mimeType);

  /// The expected size of the download in total number of bytes based on the
  /// HTTP content-length header. Returns -1 if the size is unknown.
  [propget] HRESULT ExpectedDownloadSizeInBytes(
    [out, retval] INT64* expectedDownloadSizeInBytes);

  /// The number of bytes that have been written to the download file.
  [propget] HRESULT DownloadProgressSizeInBytes([out, retval] UINT64* downloadProgressSizeInBytes);

  /// The estimated end time in ISO 8601 format.
  [propget] HRESULT EstimatedEndTime([out, retval] LPWSTR* estimatedEndTime);

>   /// The absolute path to the file, including file name. Host can change

Reviewer 4: ```suggestion
  /// The absolute path to the download file, including file name. Host can change
```

  /// Cancels the download. If canceled, the default download dialog will show
  /// that the download was canceled. Host should set the `Cancel` property from
  /// `ICoreWebView2SDownloadStartingEventArgs` if the download should be
  /// cancelled without displaying the default download dialog.
  HRESULT Cancel();

  /// Pauses the download. If paused, the default download dialog will
  /// show that the download is paused. No effect if download is already paused.
  /// Pausing a download will change the state to `COREWEBVIEW2_DOWNLOAD_STATE_INTERRUPTED`
  /// with `InterruptReason` set to `COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_PAUSED`.
  HRESULT Pause();

  /// Resumes a paused download. May also resume a download that was interrupted
  /// for another reason, if `CanResume` returns true. Resuming a download will
  /// change the state from `COREWEBVIEW2_DOWNLOAD_STATE_INTERRUPTED` to
  /// `COREWEBVIEW2_DOWNLOAD_STATE_IN_PROGRESS`.
  HRESULT Resume();

  /// Returns whether user has paused the download.
  [propget] HRESULT IsPaused([out, retval] BOOL* isPaused);

  /// Returns true if an interrupted download can be resumed. Downloads with

>   /// the following interrupt reasons may be auto-resumed:

Reviewer 4: ```suggestion
  /// the following interrupt reasons may automatically resume without you calling any methods:
```

    existing page(s) or spec documentation.

    For example, this section is a place to explain why you're adding this
    API rather than modifying an existing API.

    For example, this is a place to provide a brief explanation of some dependent
    area, just explanation enough to understand this new API, rather than telling
    the reader "go read 100 pages of background information posted at ...".
-->
We are redesigning the download experience by providing a DownloadStarting event, and exposing a Download object with all associated download metadata. With this you will be able to block downloads, save to a different path, and have access to the required metadata to build your own download UI.

In this document we describe the updated API. We'd appreciate your feedback.

# Description

There are 3 parts to this API: the CoreWebView2.DownloadStarting event that can be used to find out about or cancel new downloads, the Download class available from the DownloadStarting event that has up to date information on a download as it progresses, and the CoreWebView2Settings.DefaultDownloadDialog property that can be used to disable the default download UI.

> 1. DownloadStarting [Event] - used to intercept a download.

Reviewer 4: Please remove this 1,2,3. We have summary of the 3 above and then details in the API Details section at the bottom.
